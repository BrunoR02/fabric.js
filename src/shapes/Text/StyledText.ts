import type { ObjectEvents } from '../../EventTypeDefs';
import type {
  FabricObjectProps,
  SerializedObjectProps,
  TProps,
} from '../Object/types';
import { FabricObject } from '../Object/FabricObject';
import { styleProperties } from './constants';
import type { StylePropertiesType } from './constants';
import type { Text } from './Text';
import { pick } from '../../util';

export type CompleteTextStyleDeclaration = Pick<Text, StylePropertiesType>;

export type TextStyleDeclaration = Partial<CompleteTextStyleDeclaration>;

export type TextStyle = {
  [line: number | string]: { [char: number | string]: TextStyleDeclaration };
};

export abstract class StyledText<
  Props extends TProps<FabricObjectProps> = Partial<FabricObjectProps>,
  SProps extends SerializedObjectProps = SerializedObjectProps,
  EventSpec extends ObjectEvents = ObjectEvents
> extends FabricObject<Props, SProps, EventSpec> {
  /**
   * correlates to unwrapped lines
   */
  declare abstract styles: TextStyle;
  protected declare abstract _textLines: string[][];
  protected declare _forceClearCache: boolean;
  static _styleProperties: Readonly<StylePropertiesType[]> = styleProperties;
  abstract getStyleCursorPosition(
    index: number,
    text?: string[]
  ): {
    charIndex: number;
    lineIndex: number;
  };

  /**
   * Returns true if object has no styling or no styling in a line
   * @param {Number} lineIndex , lineIndex is on wrapped lines.
   * @return {Boolean}
   */
  isEmptyStyles(lineIndex?: number): boolean {
    if (!this.styles) {
      return true;
    }
    if (typeof lineIndex !== 'undefined' && !this.styles[lineIndex]) {
      return true;
    }
    const obj =
      typeof lineIndex === 'undefined'
        ? this.styles
        : { line: this.styles[lineIndex] };
    for (const p1 in obj) {
      for (const p2 in obj[p1]) {
        // eslint-disable-next-line no-unused-vars
        for (const p3 in obj[p1][p2]) {
          return false;
        }
      }
    }
    return true;
  }

  /**
   * Returns true if object has a style property or has it ina specified line
   * This function is used to detect if a text will use a particular property or not.
   * @param {String} property to check for
   * @param {Number} lineIndex to check the style on
   * @return {Boolean}
   */
  styleHas(property: keyof TextStyleDeclaration, lineIndex?: number): boolean {
    if (!this.styles) {
      return false;
    }
    if (typeof lineIndex !== 'undefined' && !this.styles[lineIndex]) {
      return false;
    }
    const obj =
      typeof lineIndex === 'undefined'
        ? this.styles
        : { 0: this.styles[lineIndex] };
    // eslint-disable-next-line
    for (const p1 in obj) {
      // eslint-disable-next-line
      for (const p2 in obj[p1]) {
        if (typeof obj[p1][p2][property] !== 'undefined') {
          return true;
        }
      }
    }
    return false;
  }

  /**
   * Gets style of a current selection/cursor (at the start position)
   * @param {Number} startIndex Start index to get styles at
   * @param {Number} endIndex End index to get styles at, if not specified startIndex + 1
   * @param {Boolean} [complete] get full style or not
   * @return {Array} styles an array with one, zero or more Style objects
   */
  getSelectionStyles(
    startIndex: number,
    endIndex?: number,
    complete?: boolean
  ) {
    const styles: TextStyleDeclaration[] = [];
    for (let i = startIndex; i < (endIndex || startIndex); i++) {
      styles.push(this.getStyleAtPosition(i, complete));
    }
    return styles;
  }

  /**
   * Gets style of a current selection/cursor position
   * @param {Number} position  to get styles at
   * @param {Boolean} [complete] full style if true
   * @return {Object} style Style object at a specified index
   * @private
   */
  getStyleAtPosition<
    T extends boolean = false,
    R = T extends true ? CompleteTextStyleDeclaration : TextStyleDeclaration
  >(position: number, complete?: T): R {
    const { lineIndex, charIndex } = this.getStyleCursorPosition(position);
    return {
      ...(complete
        ? // @ts-expect-error readonly
          pick(this, (this.constructor as typeof StyledText)._styleProperties)
        : {}),
      ...(this.styles[lineIndex]?.[charIndex] || {}),
    } as R;
  }

  /**
   * Sets style of a current selection, if no selection exist, do not set anything.
   * @param {Object} styles Styles object
   * @param {Number} startIndex Start index to get styles at
   * @param {Number} [endIndex] End index to get styles at, if not specified startIndex + 1
   */
  setSelectionStyles(
    style: TextStyleDeclaration,
    startIndex: number,
    endIndex?: number
  ) {
    for (let i = startIndex; i < (endIndex || startIndex); i++) {
      const { lineIndex, charIndex } = this.getStyleCursorPosition(i);
      (this.styles[lineIndex] || (this.styles[lineIndex] = {}))[charIndex] = {
        ...style,
      };
    }
    /* not included in _extendStyles to avoid clearing cache more than once */
    this._forceClearCache = true;
  }

  /**
   * get the reference, not a clone, of the style object for a given character,
   * if not style is set for a pre det
   * @param {Number} lineIndex
   * @param {Number} charIndex
   * @return {Object} style object a REFERENCE to the existing one or a new empty object
   */
  _getStyleDeclaration(
    lineIndex: number,
    charIndex: number
  ): TextStyleDeclaration {
    const lineStyle = this.styles && this.styles[lineIndex];
    return lineStyle ? lineStyle[charIndex] ?? {} : {};
  }

  /**
   * return a new object that contains all the style property for a character
   * the object returned is newly created
   * @param {Number} lineIndex of the line where the character is
   * @param {Number} charIndex position of the character on the line
   * @return {Object} style object
   */
  getCompleteStyleDeclaration(
    lineIndex: number,
    charIndex: number
  ): CompleteTextStyleDeclaration {
    return {
      // @ts-expect-error readonly
      ...pick(this, (this.constructor as typeof StyledText)._styleProperties),
      ...this._getStyleDeclaration(lineIndex, charIndex),
    } as CompleteTextStyleDeclaration;
  }
}
