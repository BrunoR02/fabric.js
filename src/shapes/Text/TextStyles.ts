import { pick } from '../../util';
import { pickBy } from '../../util/misc/pick';
import { styleProperties, type StylePropertiesType } from './constants';
import type { Text } from './Text';

export type CompleteTextStyleDeclaration = Pick<Text, StylePropertiesType>;

export type TextStyleDeclaration = Partial<CompleteTextStyleDeclaration>;

export type LegacyTextStyle = {
  [line: number | string]: { [char: number | string]: TextStyleDeclaration };
};

export type TextStyleJSON = {
  start: number;
  end: number;
  style: TextStyleDeclaration;
};

type PositionOrOffset =
  | { lineIndex: number; charIndex: number; offset?: never }
  | { lineIndex?: never; charIndex?: never; offset: number };

type StyleOptions =
  | {
      complete?: boolean;
      uniq?: never;
    }
  | {
      complete?: never;
      uniq?: boolean;
    };

/**
 * @param {Object} prevStyle first style to compare
 * @param {Object} thisStyle second style to compare
 * @param {boolean} forTextSpans whether to check overline, underline, and line-through properties
 * @return {boolean} true if the style changed
 */
export const hasStyleChanged = (
  prevStyle: TextStyleDeclaration,
  thisStyle: TextStyleDeclaration,
  forTextSpans = false
) =>
  prevStyle.fill !== thisStyle.fill ||
  prevStyle.stroke !== thisStyle.stroke ||
  prevStyle.strokeWidth !== thisStyle.strokeWidth ||
  prevStyle.fontSize !== thisStyle.fontSize ||
  prevStyle.fontFamily !== thisStyle.fontFamily ||
  prevStyle.fontWeight !== thisStyle.fontWeight ||
  prevStyle.fontStyle !== thisStyle.fontStyle ||
  prevStyle.textBackgroundColor !== thisStyle.textBackgroundColor ||
  prevStyle.deltaY !== thisStyle.deltaY ||
  (forTextSpans &&
    (prevStyle.overline !== thisStyle.overline ||
      prevStyle.underline !== thisStyle.underline ||
      prevStyle.linethrough !== thisStyle.linethrough));

export class TextStyles {
  protected styles: TextStyleDeclaration[];

  constructor(
    readonly target: Text,
    styles?: TextStyleDeclaration[] | TextStyleJSON[] | LegacyTextStyle
  ) {
    const len = this.target._text.length;
    if (Array.isArray(styles) && styles.length === len) {
      this.styles = styles as TextStyleDeclaration[];
    } else {
      // backward compatibility
      this.styles = new Array(len).fill({});
      if (Array.isArray(styles)) {
        // compact style array
        (styles as TextStyleJSON[]).forEach(({ start, end, style }) => {
          for (let index = start; index < end; index++) {
            this.styles[index] = { ...style };
          }
        });
      } else if (typeof styles === 'object') {
        // legacy styles
        for (const lineIndex in styles) {
          for (const charIndex in styles[lineIndex]) {
            this.set({
              lineIndex: Number(lineIndex),
              charIndex: Number(charIndex),
              style: { ...styles[lineIndex][charIndex] },
            });
          }
        }
      }
    }
  }

  protected getLines() {
    return this.target._textLines;
  }

  positionToOffset(lineIndex: number, charIndex = 0) {
    const lines = this.getLines();
    let total = 0;
    for (let index = 0; index < lineIndex; index++) {
      total += lines[index].length;
    }
    return total + charIndex;
  }

  resolveOffset({ lineIndex, charIndex, offset }: PositionOrOffset) {
    return typeof offset === 'number'
      ? offset
      : this.positionToOffset(lineIndex, charIndex);
  }

  slice(
    start: number,
    end?: number,
    { uniq = false, complete = false }: StyleOptions = {}
  ) {
    // @ts-expect-error readonly
    const upstream = complete ? pick(this.target, styleProperties) : {};
    return this.styles.slice(start, end).map(
      (style) =>
        ({
          ...upstream,
          ...(uniq && !complete
            ? pickBy(style, (v, k) => this.target[k] !== v)
            : style),
        } as TextStyleDeclaration)
    );
  }

  get({
    uniq = false,
    complete = false,
    ...position
  }: PositionOrOffset & StyleOptions) {
    // @ts-expect-error readonly
    const upstream = complete ? pick(this.target, styleProperties) : {};
    const style = this.styles[this.resolveOffset(position)];
    return {
      ...upstream,
      ...(uniq && !complete
        ? pickBy(style, (v, k) => this.target[k] !== v)
        : style),
    } as TextStyleDeclaration;
  }

  value({
    key,
    uniq = false,
    complete = false,
    ...position
  }: PositionOrOffset & StyleOptions & { key: keyof TextStyleDeclaration }) {
    const value = this.styles[this.resolveOffset(position)][key];
    return complete
      ? value ?? this.target[key]
      : !uniq || value !== this.target[key]
      ? value
      : undefined;
  }

  set({
    style,
    repeatCount,
    ...position
  }: PositionOrOffset & { repeatCount?: number; style: TextStyleDeclaration }) {
    const offset = this.resolveOffset(position);
    this.styles[offset] = style;
    for (let index = 0; index < offset + (repeatCount || 1); index++) {
      this.styles[offset] = style;
    }
  }

  clear(position: PositionOrOffset) {
    this.set({ ...position, style: {} });
  }

  splice(
    position: PositionOrOffset,
    deleteCount: number,
    styles: TextStyleDeclaration[] = []
  ) {
    this.styles.splice(this.resolveOffset(position), deleteCount, ...styles);
  }

  reset(styles: TextStyleDeclaration[] = []) {
    this.styles = styles;
  }

  has({
    lineIndex,
    charIndex,
    offset,
    key,
  }:
    | {
        lineIndex?: never;
        charIndex?: never;
        offset?: number;
        key?: keyof TextStyleDeclaration;
      }
    | {
        lineIndex: number;
        charIndex?: number;
        offset?: never;
        key?: keyof TextStyleDeclaration;
      } = {}): boolean {
    let slice: TextStyleDeclaration[];
    if (typeof offset === 'number') {
      slice = [this.styles[offset]];
    } else if (typeof lineIndex === 'number') {
      const start = this.positionToOffset(lineIndex, charIndex || 0);
      slice = this.styles.slice(
        start,
        start + typeof charIndex === 'number'
          ? start + 1
          : this.getLines()[lineIndex].length
      );
    } else {
      slice = this.styles;
    }

    return slice.some((value) =>
      !!value && key ? Object.hasOwn(value, key) : Object.keys(value).length > 0
    );
  }

  uniq() {
    return this.styles.map((value) =>
      pickBy(value, (v, k) => this.target[k] !== v)
    );
  }

  uniqSelf() {
    this.styles = this.uniq();
  }

  /**
   * Returns the array form of a text object's inline styles property with styles grouped in ranges
   * rather than per character. This format is less verbose, and is better suited for storage
   * so it is used in serialization (not during runtime).
   * @return {TextStyleJSON[]}
   */
  toJSON() {
    const textLines = this.getLines(),
      stylesArray: TextStyleJSON[] = [];
    let charIndex = -1,
      prevStyle = {};

    //loop through each textLine
    for (let i = 0; i < textLines.length; i++) {
      if (!this.has({ lineIndex: i })) {
        //no styles exist for this line, so add the line's length to the charIndex total and reset prevStyle
        charIndex += textLines[i].length;
        prevStyle = {};
        continue;
      }
      //loop through each character of the current line
      for (let c = 0; c < textLines[i].length; c++) {
        charIndex++;
        const thisStyle = this.get({ lineIndex: i, charIndex: c, uniq: true });
        //check if style exists for this character
        if (thisStyle && Object.keys(thisStyle).length > 0) {
          if (hasStyleChanged(prevStyle, thisStyle, true)) {
            stylesArray.push({
              start: charIndex,
              end: charIndex + 1,
              style: thisStyle,
            });
          } else {
            //if style is the same as previous character, increase end index
            stylesArray[stylesArray.length - 1].end++;
          }
        }
        prevStyle = thisStyle || {};
      }
    }
    return stylesArray;
  }
}
