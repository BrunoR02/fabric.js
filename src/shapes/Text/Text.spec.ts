import { Text } from './Text';

describe('Text', () => {
  it('toObject', async () => {
    expect(new Text('text').toObject()).toMatchSnapshot();
  });
  it('fromObject', async () => {
    expect((await Text.fromObject({ text: 'text' })).toObject()).toEqual(
      new Text('text').toObject()
    );
  });
});
