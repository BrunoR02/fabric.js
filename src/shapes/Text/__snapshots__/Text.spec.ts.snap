// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text toObject 1`] = `
{
  "angle": 0,
  "backgroundColor": "",
  "charSpacing": 0,
  "direction": "ltr",
  "fill": "rgb(0,0,0)",
  "fillRule": "nonzero",
  "flipX": false,
  "flipY": false,
  "fontFamily": "Times New Roman",
  "fontSize": 40,
  "fontStyle": "normal",
  "fontWeight": "normal",
  "globalCompositeOperation": "source-over",
  "height": 45.2,
  "left": 0,
  "lineHeight": 1.16,
  "linethrough": false,
  "opacity": 1,
  "originX": "left",
  "originY": "top",
  "overline": false,
  "paintFirst": "fill",
  "path": undefined,
  "pathAlign": "baseline",
  "pathSide": "left",
  "pathStartOffset": 0,
  "scaleX": 1,
  "scaleY": 1,
  "shadow": null,
  "skewX": 0,
  "skewY": 0,
  "stroke": null,
  "strokeDashArray": null,
  "strokeDashOffset": 0,
  "strokeLineCap": "butt",
  "strokeLineJoin": "miter",
  "strokeMiterLimit": 4,
  "strokeUniform": false,
  "strokeWidth": 1,
  "styles": [],
  "text": "text",
  "textAlign": "left",
  "textBackgroundColor": "",
  "top": 0,
  "type": "Text",
  "underline": false,
  "version": "6.0.0-beta10",
  "visible": true,
  "width": 2,
}
`;
